packages/flame_studio/lib/src/widgets/panels/hierarchy_view.dart

```diff
@@ -0,0 +1,221 @@
+import 'package:flame/components.dart';
+import 'package:flame_studio/src/core/component_tree.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class HierarchyView extends ConsumerStatefulWidget {
+  const HierarchyView({super.key});
+
+  @override
+  HierarchyViewState createState() => HierarchyViewState();
+}
+
+class HierarchyViewState extends ConsumerState<ConsumerStatefulWidget> {
+  Component? selectedComponent;
+  Set<Component> expandedComponents = {};
+
+  @override
+  Widget build(BuildContext context) {
+    final componentTree = ref.watch(componentTreeProvider).root;
+    if (componentTree == null) {
+      return Container();
+    }
+
+    final listItems = <Widget>[];
+    _buildList(componentTree, 0, listItems);
```

Comment:
It's not recommended to build widgets with functions like this afaik:
https://github.com/flutter/flutter/issues/19269

Reply:
It would be cleaner to just create a `StatelessWidget` that takes in the the `componentTree` here instead.

Reply:
`_buildList` doesn't create a widget, it populates `listItems` with widgets for each line item in the list. And it has to be a function, because it is called recursively in order to traverse the tree (of course, recursion can be replaced with a stack of iterators, but it's much cleaner via the recursion).

Reply:
Right, gotcha! This could recursively build up a model that is then passed into a widget that creates the visual parts.
The current implementation doesn't feel very flutter-esque at least. Wdyt @flame-engine/flame-admin?

Reply:
`_buildList` does creates widgets, I mean, it is populating a list with widgets that it creates.

It is not clear to me it needs to be a function, it could just be a Widget that once it identifies that it has to go into a node, it can just add "itself" as the part of its own children, forwarding the children of that node to the new instance, doing a recursion the same way

Reply:
I can make each item in this list a separate widget though.

Reply:
So the plan is to make them widgets then? :)

Reply:
Already did